### PR TITLE
docs: how to support a new model tutorial

### PR DESCRIPTION
**Pull request by jonathlela**

### CODE REVIEW

**Comment by pommedeterresautee** [comment]
Pr name should start by docs: ...

File: docs/get-started/support-new-model.md [hunk 1]
```diff
@@ -0,0 +1,548 @@
...
+
+``` { .py }
+def _compiler(gm: torch.fx.GraphModule, example_inputs: List[torch.Tensor]):
+    dynamo_backend_ofi(gm)
+    return cuda_graphs_wrapper(gm, example_inputs, pool=_pool)
+```
+This backend combines two steps:
+
+- First one is to apply the graph replacements
+- Second one is to use [CUDA graphs](https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/)
```
Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:22]: As we don't elaborate may be just say it is to remove python overhead, etc.

File: docs/get-started/support-new-model.md [hunk 2]
```diff
@@ -0,0 +1,548 @@
...
+
+We won't elaborate the second step and focus on the first one that actually does the graph replacements.
+
+### Replace part of the FX graph
+
+First of all, a few word about [Torch FX](https://pytorch.org/docs/1.13/fx.html).
```
Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:28]: Words

File: docs/get-started/support-new-model.md [hunk 3]
```diff
@@ -0,0 +1,548 @@
...
+Torch FX is a torch module to module transformation toolkit. It can trace an torch module (or a function) execution. All the operation are then recorded into a list of nodes that are represented as an FX Graph. From this graph, Torch FX generates python code matching the graph's semantic. Both graph and python code are accessible from the Torch FX GraphModule, which is also a torch module instance. Playing with Torch FX allows us to modify a graph but to stay at the torch module level.
```
Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:29]: Module repeated in 1st sentence

Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:29]: The explanation seems to be very detailed. May be something related to it traces computation graph + let us manipulate it? And a special word about dynamic behavior?

File: docs/get-started/support-new-model.md [hunk 4]
```diff
@@ -0,0 +1,548 @@
...
+  F(Torch Module) -.-> C;
+  B --> C[Torch FX GraphModule];
+  C --> D(Torch FX Graph);
+  C --> E(Python Code);
+  D -. Share same semantic .- E;
+```
+
+It's possible to [directly modify the FX graph](https://pytorch.org/docs/1.13/fx.html#direct-graph-manipulation) but for this tutorial we'll rather focus on subgraph rewriting.
+
+To rewrite the Torch FX graph, kernl use the `#!python repace_pattern()` function defined in [src/kernl/utils/extended_matcher.py](https://github.com/ELS-RD/kernl/blob/1463e39aeaaa49535c4a9144c4ab96c6a7a15eca/src/kernl/utils/extended_matcher.py#L337). It's the same function `#!python repace_pattern()` of [Torch FX](https://pytorch.org/docs/1.13/fx.html#subgraph-rewriting-with-replace-pattern) but with some bugfixes (that should be integrated in PyTorch in the future).
```
Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:46]: Uses

Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:46]: There is no bug fixes, it's more another way to do things which is more focused on our needs. I let you see with @gaetansnl what it is exactly, I will wait your next version to understand myself :-)

File: docs/get-started/support-new-model.md [hunk 5]
```diff
@@ -0,0 +1,548 @@
...
+                outputs = outputs + (past_key_value,)
+            return outputs
+    ```
+
+    1.  $QK^T$
+    2.  $\frac{QK^T}{\sqrt{d_k}}$
+    3.  $\operatorname{softmax}(\frac{QK^T}{\sqrt{d_k}})$
+    4.  $\operatorname{softmax}(\frac{QK^T}{\sqrt{d_k}})V$
+
+We see that the Hugging Face implementation is close to the definition in the paper, we can actually want to find the attention pattern in this model.
```
Comment (pommedeterresautee) [inline docs/get-started/support-new-model.md:299]: From the

**Review by pommedeterresautee — COMMENTED** [review]
In general I think i learned things. May be some parts can be written in a simpler way. Also if you want I can send the markdown file to Antidote grammar corrector, I have put some comments but not sure it covers everything. Let me know.

File: docs/get-started/support-new-model.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (gaetansnl) [reply docs/get-started/support-new-model.md:29]: "Playing with Torch FX allows us to modify a graph but to stay at the torch module level." not sure to understand what you mean

File: docs/get-started/support-new-model.md [hunk 6]
```diff
@@ -0,0 +1,548 @@
...
+``` { .py }
+torch.fx.wrap(fn)
+```
+
+### Inspecting the FX Graph
+
+!!! todo
+    introduce `#!python graph_report(gm)` and `#!python gm.code`
+
+There is to way to inspect the FX Graph. One is to print the graph as a list of nodes, the other one is the python code
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:106]: are, two

File: docs/get-started/support-new-model.md [hunk 7]
```diff
@@ -0,0 +1,548 @@
...
+
+There is to way to inspect the FX Graph. One is to print the graph as a list of nodes, the other one is the python code 
+
+https://pytorch.org/docs/1.13/fx.html#torch.fx.Graph.print_tabular
+
+https://pytorch.org/docs/1.13/fx.html#torch.fx.Node
+
+## Example: replacing BERT Attention
+
+In this example, we'll see how to replace the attention part of a BERT model with kernl's optimized attention kernel.
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:114]: Uppercase Kernl

File: docs/get-started/support-new-model.md [hunk 8]
```diff
@@ -0,0 +1,548 @@
...
+<figure markdown>
+  ![Scaled Dot-Product Attention](attention_original.png){ width="150"; lazyload=true }
+  <figcaption>Scaled Dot-Product Attention</figcaption>
+</figure>
+
+This graph representation will be useful as it is this graph we'll try to replace to optimize a BERT model.
+
+### Find the Attention graph pattern
+
+For our example, we'll try to replace the attention part from the "bert-base-uncased" pre-trained model from [Hugging Face Transformers](https://huggingface.co/transformers). If we look at the [BERT implementation](https://github.com/huggingface/transformers/blob/main/src/transformers/models/bert/modeling_bert.py), we find the attention function as a torch module
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:151]: we do more than "trying" :)

File: docs/get-started/support-new-model.md [hunk 9]
```diff
@@ -0,0 +1,548 @@
...
+    truediv = matmul / 8.0
+    add_1 = truediv + attention_mask
+    softmax = torch.nn.functional.softmax(add_1, dim=-1)
+    matmul_1 = torch.matmul(softmax, v)
+    return matmul_1
+```
+
+### Replace the Attention part
+
+We now needs to add our replace function to call the optimized kernel. We can see in [kernl/model_optimization.py](https://github.com/ELS-RD/kernl/blob/9c3ec9b3c03609c20ca9a00eda7a58d4769bf47c/src/kernl/implementations/attention.py#L492) the optimized attention kernel needs in addition to `q`, `k`, `v` and `attention_mask`, the `output` and the `sm_scale` parameter.
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:462]: Instead of what in links use version tags to avoid problems in the future

File: docs/get-started/support-new-model.md [hunk 10]
```diff
@@ -0,0 +1,548 @@
...
+        matmul = torch.matmul(q, transpose)
+        truediv = matmul / 8.0
+        add_1 = truediv + attention_mask
+        softmax = torch.nn.functional.softmax(add_1, dim=-1)
+        matmul_1 = torch.matmul(softmax, v)
+        return matmul_1
+
+    def replace(q, k, attention_mask, v):
+        output = torch.empty_like(q)
+        output = attention_forward(q, k, v, output, 1 / 8.0, is_causal=False, attention_mask=attention_mask)
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:515]: maybe explain what is `is_causal` ? maybe in another doc, I don't know

File: docs/get-started/support-new-model.md [hunk 11]
```diff
@@ -0,0 +1,548 @@
...
+
+with torch.inference_mode(), torch.cuda.amp.autocast(enabled=True, dtype=torch.float16, cache_enabled=True):
+    inputs = {
+        "input_ids": torch.randint(2, 10000, shape, device="cuda", dtype=torch.long),
+        "attention_mask": torch.ones(shape, device="cuda", dtype=torch.long),
+    }
+    output = model(**inputs)
+```
+
+If we run `#!python graph_report(gm)` and `#!python gm.code` in the `#!python dynamo_backend_ofi(gm)` function, we can print the FX graph and the python code from the model computation. For our example, we'll keep the `normalize_operators` and the `remove_dropout` functions as it simplifies a bit the model's graph.
```
Comment (gaetansnl) [inline docs/get-started/support-new-model.md:323]: Are you suggesting the user should modify our code ? Maybe the tutorial should be written in such a way we can do everything without modifying our code, but I'm not sure if it's possible with current codebase

File: docs/get-started/support-new-model.md (hunk 3 above)
Comment (jonathlela) [reply docs/get-started/support-new-model.md]: Actually I've detailed a bit as I explain more how torch fx works afterwards (not in the draft version you saw)

File: docs/get-started/support-new-model.md (hunk 11 above)
Comment (jonathlela) [reply docs/get-started/support-new-model.md]: It's not really possible with the actual codebase.  I was thinking of two ways.

The first one is to write a jupyter notebook where we only overload the optimization function and play with it. But a jupyter notebook is not really easy for a first time read.

The second one was to adapt our codebase easily add / remove optimization (like a pluggin system), and one of the "plugin" only prints the graph. But it demands a lot of work on the code.

File: docs/get-started/support-new-model.md (hunk 10 above)
Comment (jonathlela) [reply docs/get-started/support-new-model.md:653]: As it's already a bit complicated, I've explained the minimal way to wrap the attention kernel. But, if it's misleading I can add a few words about it.

File: docs/get-started/support-new-model.md [hunk 12]
```diff
@@ -0,0 +1,548 @@
...
+    def pattern(w1, w2):
+        return torch.cat([w1, w2]).sum()
+
+    def replacement(w1, w2):
+        return torch.stack([w1, w2])
+
+    replace_pattern(gm, pattern, replacement)
+    ```
+
+When using a non-torch function in a subgraph, we'll have to use [Torch wrap function](https://pytorch.org/docs/1.13/fx.html#non-torch-functions) in order to appears in the FX graph but not to be traced
```
[the remark this answers is not in the document]
Comment (jonathlela) [reply docs/get-started/support-new-model.md]: I've added this slight difference in the text.

File: docs/get-started/support-new-model.md (hunk 3 above)
[the remark this answers is not in the document]
Comment (jonathlela) [reply docs/get-started/support-new-model.md]: 1. To explain more in depth how torch fx works pour replacements, I use the `symbolic_trace` function. We don't use it directly on our code but the replace_pattern function use it for the input.

2. I try to emphasize that we replace graphs but in the end, we're still having torch module objects. I slightly rephrased it.

**Comment by white-gorilla** [comment]
I don't understand the details of the documentation, but from what I understand, I find it very clear. The diagrams and examples are self-explanatory.

On the other hand, the organization of the documentation must be addressed.

Here we have a first tutorial in the `get started` section.
I think we have two options.

- Either this tutorial is part of a `get started`, then we can leave it in place as long as we make it part of the topic, with a `Tutorial` sub-section if we have more than one.

Ex. :
```
# Page tree
nav:
  - Home: index.md
  - Get Started:
      - Try kernl: get-started/start.md
      - Tutorial:
        - How to support a new model: get-started/support-new-model.md
```
Capture d’écran 2022-12-01 à 11 49 20

- Either we create a dedicated `Tutorial` topic or section, as we will have enough tutorials.

In this case, in my opinion, it is necessary to delete a topic and it is necessary to move the documentation to the right directory (the old directory renamed to `tutorial`).

```
# Page tree
nav:
  - Home: index.md
  - Get Started:
      - Try kernl: get-started/start.md
  - Tutorial:
      - How to support a new model: tutorial/support-new-model.md
```

Capture d’écran 2022-12-01 à 11 55 20

**Review by white-gorilla — CHANGES_REQUESTED** [review]
see: https://github.com/ELS-RD/kernl/pull/181#issuecomment-1333583277

**Review by white-gorilla — APPROVED** [review]
Is it possible to put a level 2 title in the Get started document to have a table of contents on the right? Otherwise it makes an empty column.

Ex.
Capture d’écran 2022-12-07 à 18 24 54

File: docs/how-to-guides/support-new-model.md [hunk 13]
```diff
@@ -0,0 +1,688 @@
...
+- First one is to apply the graph replacements
+- Second one is to use [CUDA graphs](https://pytorch.org/blog/accelerating-pytorch-with-cuda-graphs/)
+
+The second step eliminates most of the CPU overhead but we won't elaborate on this and focus on the first one that does the graph replacements.
+
+
+### Inspecting the FX Graph
+
+First, a few words about [Torch FX](https://pytorch.org/docs/1.13/fx.html).
+Torch FX is a torch module to module transformation toolkit. It can trace a torch module (or a function) execution. All the operations are then recorded into a list of nodes that are represented as an FX Graph. From this graph, Torch FX generates python code matching the graph's semantics. Both graph and python code are accessible from the Torch FX GraphModule, which is also a torch module instance. Torch FX allows us to play with FX graphs but stay at the torch module level.
```
Comment (pommedeterresautee) [inline docs/how-to-guides/support-new-model.md:27]: into a list of nodes that are represented as an FX Graph -> into a graph of nodes

File: docs/how-to-guides/support-new-model.md [hunk 14]
```diff
@@ -0,0 +1,688 @@
+# How to support a new model
+
+## How does Kernl optimize a model
+
+### Overview
+
+To optimize a model, Kernl uses [TorchDynamo](https://github.com/pytorch/torchdynamo) JIT compiler and provides a custom backend where we replace part of the [Torch FX](https://pytorch.org/docs/1.13/fx.html) graph with optimized kernels.
```
Comment (pommedeterresautee) [inline docs/how-to-guides/support-new-model.md:7]: FX is not yet defined, I would just say we replace a part of the model with our own kernels

File: docs/how-to-guides/support-new-model.md [hunk 15]
```diff
@@ -0,0 +1,688 @@
...
+
+First, a few words about [Torch FX](https://pytorch.org/docs/1.13/fx.html).
+Torch FX is a torch module to module transformation toolkit. It can trace a torch module (or a function) execution. All the operations are then recorded into a list of nodes that are represented as an FX Graph. From this graph, Torch FX generates python code matching the graph's semantics. Both graph and python code are accessible from the Torch FX GraphModule, which is also a torch module instance. Torch FX allows us to play with FX graphs but stay at the torch module level.
+
+<figure markdown>
+  ![Torch FX](torchfx.drawio.svg){ lazyload=true }
+  <figcaption>Torch FX</figcaption>
+</figure>
+
+With a FX GraphModule, we can inspect both the FX Graph and the generated code, with `.graph` and `.code` respectively. For better readability, one may want to use `#!python graph_report(gm)` that print the FX Graph in a tabular way (similarly to [Torch FX `print_tabular` method](https://pytorch.org/docs/1.13/fx.html#torch.fx.Graph.print_tabular)).
```
Comment (pommedeterresautee) [inline docs/how-to-guides/support-new-model.md:34]: can you cite the official way through `config` variables?
https://github.com/pytorch/pytorch/blob/master/torch/_dynamo/config.py#L39

**Review by pommedeterresautee — APPROVED** [review]
I made some small remarks but it seems very good to me.

File: docs/how-to-guides/support-new-model.md (hunk 13 above)
Comment (pommedeterresautee) [reply docs/how-to-guides/support-new-model.md]: Addressed in 7722ebd4ea1d643bfc1796d598711f4b56142692

File: docs/how-to-guides/support-new-model.md (hunk 15 above)
Comment (pommedeterresautee) [reply docs/how-to-guides/support-new-model.md]: addressed in 7722ebd4ea1d643bfc1796d598711f4b56142692

File: docs/how-to-guides/support-new-model.md (hunk 14 above)
Comment (pommedeterresautee) [reply docs/how-to-guides/support-new-model.md:7]: not very important.
